File: iosApp/iosApp/Firebase/FirebaseDatabase.swift
import FirebaseDatabase

final class FirebaseDatabase {
    private static let database = Database.database()
    private static let reference = database.reference()

    func goOffline() {
        FirebaseDatabase.database.goOffline()
    }
    func goOnline() {
        FirebaseDatabase.database.goOnline()
    }
    func getData(_ child: String, callback: @escaping (Any?) -> Void) {
        FirebaseDatabase.reference.child(child).getData { error, result in
            if error != nil {
                callback(nil)
            }
            else {
                guard let result = result else {
                    callback(nil)
                    return
                }
                callback(result.valueInExportFormat())
            }
        }
    }
    func getDataRealtime(_ child: String, callback: @escaping (Any?) -> Void) {
        FirebaseDatabase.reference.child(child).observe(.value) { result in
            if (!result.exists()) {
                callback(nil)
            }
            else {
                callback(result.valueInExportFormat())
            }
        }
    }
    func getLastData(_ child: String, toLast: UInt = 1, callback: @escaping (Any?) -> Void) {
        FirebaseDatabase.reference.child(child).queryLimited(toLast: toLast).getData { error, result in
            if error != nil {
                callback(nil)
            }
            else {
                guard let result = result else {
                    callback(nil)
                    return
                }
                callback(result.valueInExportFormat())
            }
        }
    }
    func getLastDataRealtime(_ child: String, toLast: UInt = 1, callback: @escaping (Any?) -> Void) {
        FirebaseDatabase.reference.child(child).queryLimited(toLast: toLast).observe(.value) { result in
            if (!result.exists()) {
                callback(nil)
            }
            else {
                callback(result.valueInExportFormat())
            }
        }
    }
    func getLastData(_ child: String, atValue: String?, callback: @escaping (Any?) -> Void) {
        if (atValue == nil) {
            getData(child, callback: {
                callback($0)
            })
        }
        FirebaseDatabase.reference.child(child).queryOrderedByKey().queryStarting(atValue: atValue).getData { error, result in
            if error != nil {
                callback(nil)
            }
            else {
                guard let result = result else {
                    callback(nil)
                    return
                }
                callback(result.valueInExportFormat())
            }
        }
    }
    func getLastDataRealtime(_ child: String, atValue: String?, callback: @escaping (Any?) -> Void) {
        if (atValue == nil) {
            getDataRealtime(child, callback: {
                callback($0)
            })
        }
        else {
            FirebaseDatabase.reference.child(child).queryOrderedByKey().queryStarting(atValue: atValue).observe(.value) { result in
                if (!result.exists()) {
                    callback(nil)
                }
                else {
                    callback(result.valueInExportFormat())
                }
            }
        }
    }
    func setData(_ child: String, data: [AnyHashable : Any], callback: @escaping (String?) -> Void) {
        FirebaseDatabase.reference.child(child).updateChildValues(data) { error, _ in
            if let error = error {
                callback(error.localizedDescription)
            }
            else {
                callback(nil)
            }
        }
    }
    func deleteData(_ child: String, callback: @escaping (String?) -> Void) {
        FirebaseDatabase.reference.child(child).removeValue { error, _ in
            if let error = error {
                callback(error.localizedDescription)
            }
            else {
                callback(nil)
            }
        }
    }
}

File: androidApp/src/main/java/com/cistus/hunter/android/firebase/FirebaseDatabase.kt
package com.cistus.hunter.android.firebase

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.ktx.database
import com.google.firebase.ktx.Firebase

open class FirebaseDatabase {
    companion object {
        private val database = Firebase.database
        private val reference = database.reference
    }

    fun getData(child: String, callback: (Any?) -> Unit) {
        reference.child(child).get()
            .addOnSuccessListener {
                if (it.exists())
                    callback(it.getValue(true))
                else
                    callback(null)
            }
            .addOnFailureListener { callback(null) }
    }
    fun getDataRealtime(child: String, callback: (Any?) -> Unit) {
        reference.child(child).addValueEventListener(object : ValueEventListener {
            override fun onDataChange(snapshot: DataSnapshot) {
                if (snapshot.exists())
                    callback(snapshot.getValue(true))
                else
                    callback(null)
            }
            override fun onCancelled(error: DatabaseError) {
                callback(null)
            }
        })
    }
    fun getLastData(child: String, toLast: Int = 1, callback: (Any?) -> Unit) {
        reference.child(child).orderByKey().limitToLast(toLast)
            .addListenerForSingleValueEvent(object: ValueEventListener{
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (!snapshot.exists())
                        callback(null)
                    else
                        callback(snapshot.getValue(true))
                }
                override fun onCancelled(error: DatabaseError) {
                    callback(null)
                }
            })
    }
    fun getLastData(child: String, atValue: String?, callback: (Any?) -> Unit) {
        if (atValue == null) {
            getData(child) {
                callback(it)
            }
            return
        }
        reference.child(child).orderByKey().startAt(atValue)
            .addListenerForSingleValueEvent(object: ValueEventListener{
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (!snapshot.exists())
                        callback(null)
                    else
                        callback(snapshot.getValue(true))
                }
                override fun onCancelled(error: DatabaseError) {
                    callback(null)
                }
            })
    }
    fun getLastDataRealtime(child: String, toLast: Int = 1, callback: (Any?) -> Unit) {
        reference.child(child).orderByKey().limitToLast(toLast)
            .addValueEventListener(object: ValueEventListener{
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (!snapshot.exists())
                        callback(null)
                    else
                        callback(snapshot.getValue(true))
                }
                override fun onCancelled(error: DatabaseError) {
                    callback(null)
                }
            })
    }
    fun getLastDataRealtime(child: String, atValue: String?, callback: (Any?) -> Unit) {
        if (atValue == null) {
            getDataRealtime(child) {
                callback(it)
            }
            return
        }
        reference.child(child).orderByKey().startAt(atValue)
            .addValueEventListener(object: ValueEventListener{
                override fun onDataChange(snapshot: DataSnapshot) {
                    if (!snapshot.exists())
                        callback(null)
                    else
                        callback(snapshot.getValue(true))
                }
                override fun onCancelled(error: DatabaseError) {
                    callback(null)
                }
            })
    }
    fun setData(child: String, data: Map<String, Any>, callback: (String?) -> Unit) {
        reference.child(child).updateChildren(data)
            .addOnSuccessListener { callback(null) }
            .addOnFailureListener { callback(it.localizedMessage) }
    }
    fun deleteData(child: String, callback: (String?) -> Unit) {
        reference.child(child).removeValue()
            .addOnSuccessListener { callback(null) }
            .addOnFailureListener { callback(it.localizedMessage) }
    }
}